src/Nethermind/Nethermind.Optimism/CL/Derivation/SystemConfigDeriver.cs

```diff
@@ -0,0 +1,162 @@
+
+            foreach (var log in receipt.Logs ?? [])
+            {
+                if (log.Address == systemConfigProxy && log.Topics.Length > 0 && log.Topics[0] == SystemConfigUpdate.EventABIHash)
+                {
+                    config = UpdateSystemConfigFromLogEvent(config, log);
+                }
+            }
+        }
+
+        return config;
+    }
+
+    private SystemConfig UpdateSystemConfigFromLogEvent(SystemConfig systemConfig, LogEntryForRpc log)
+    {
+        if (log.Topics.Length != 3) throw new ArgumentException($"Expected 3 event topics (event identity, indexed version, indexed updateType), got {log.Topics.Length}");
+        if (log.Topics[0] != SystemConfigUpdate.EventABIHash) throw new ArgumentException($"Invalid {nameof(SystemConfig)} update event: {log.Topics[0]}, expected {SystemConfigUpdate.EventABIHash}");
+        if (log.Topics[1] != SystemConfigUpdate.EventVersion0) throw new ArgumentException($"Unrecognized {nameof(SystemConfig)} update event version: {log.Topics[1]}");
+
+        var updateType = log.Topics[2];
+
+        if (updateType == SystemConfigUpdate.Batcher)
+        {
+            var signature = new AbiSignature(nameof(SystemConfigUpdate.Batcher), AbiType.UInt64, AbiType.UInt64, AbiType.Address);
+            object[] decoded = AbiEncoder.Instance.Decode(AbiEncodingStyle.None, signature, log.Data);
+
+            if ((UInt64)decoded[0] != 32) throw new FormatException("Invalid pointer field");
+            if ((UInt64)decoded[1] != 32) throw new FormatException("Invalid length field");
+            var address = (Address)decoded[2];
+
+            systemConfig = systemConfig with
+            {
+                BatcherAddress = address
+            };
+        }
+        else if (updateType == SystemConfigUpdate.FeeScalars)
+        {
+            var signature = new AbiSignature(nameof(SystemConfigUpdate.FeeScalars), AbiType.UInt64, AbiType.UInt64, AbiType.Bytes32);
+            object[] decoded = AbiEncoder.Instance.Decode(AbiEncodingStyle.None, signature, log.Data);
+
+            if ((UInt64)decoded[0] != 32) throw new FormatException("Invalid pointer field");
+            if ((UInt64)decoded[1] != 32) throw new FormatException("Invalid length field");
+            var data = (byte[])decoded[2];
+            switch (data[0])
+            {
+                case 0:
+                    {
+                        break;
+                    }
+                case 1:
+                    {
+                        if (!data[1..24].IsZero())
+                            return systemConfig; // ignore
+                        break;
+                    }
+                default:
+                    {
+                        // ignore invalid type
+                        return systemConfig;
+                    }
+            }
+            systemConfig = systemConfig with
+            {
+                Scalar = data,
+            };
```

Comment:
Merge all code into a single `switch` expression:

```suggestion
            systemConfig = data[0] switch
            {
                0 => systemConfig with { Scalar = data },
                1 when data[1..24].IsZero() => systemConfig with { Scalar = data },
                _ => systemConfig // ignore all other cases
            };
```